react/OrderItems.tsx

```diff
@@ -69,164 +76,308 @@ const maybeUpdateTotalizers = (
-        ...maybeUpdateTotalizers(
-          totalizers,
-          orderFormValue,
-          items[index],
-          newItem
+      setOrderForm((prevOrderForm: OrderForm) => ({
+        ...prevOrderForm,
+        items: [...orderFormItemsRef.current, ...orderFormItems],
+        totalizers: orderFormItems.reduce(
+          addToTotalizers,
+          prevOrderForm.totalizers ?? []
         ),
-        items: updatedList,
-      })
-    },
-    [items, totalizers, orderFormValue, setOrderForm]
-  )
+        value:
+          prevOrderForm.value +
+          orderFormItems.reduce(
+            (total, item) => total + item.sellingPrice * item.quantity,
+            0
+          ),
+      }))
 
-  const mutationTask = useCallback(
-    (items: Partial<Item>[]) => async () => {
-      const {
-        data: { updateItems: newOrderForm },
-      } = await updateItems({
+      pushLocalOrderQueue({
+        type: LocalOrderTaskType.ADD_MUTATION,
         variables: {
-          orderItems: items,
+          items: mutationInput,
         },
       })
 
-      return newOrderForm
+      enqueueTask(() => {
+        return mutateAddItem({ variables: { items: mutationInput } })
+          .then(({ data }) => data!.addToCart)
+          .then(orderForm => {
+            // update the uniqueId of the items that were
+            // added locally with the value from the server
+            setOrderForm((prevOrderForm: OrderForm) => ({
+              ...prevOrderForm,
+              items: prevOrderForm.items.map(item => {
+                const inputIndex = mutationInput.findIndex(
+                  inputItem => inputItem.id === +item.id
+                )
+
+                if (inputIndex === -1) {
+                  // this item wasn't part of the initial mutation, skip it
+                  return item
+                }
+
+                const updatedItem = orderForm.items.find(
+                  updatedItem => updatedItem.id === item.id
+                )!
+
+                const fakeUniqueId = item.uniqueId
+
+                // update all mutations in the queue that referenced
+                // this item with it's fake `uniqueId`
+                updateLocalQueueItemIds({
+                  fakeUniqueId,
+                  uniqueId: updatedItem.uniqueId,
+                })
+
+                return {
+                  ...item,
+                  uniqueId: updatedItem.uniqueId,
+                }
+              }),
+            }))
+
+            return orderForm
+          })
+      })
     },
-    [updateItems]
+    [enqueueTask, mutateAddItem, setOrderForm]
   )
 
   const updateQuantity = useCallback(
-    async (props: Partial<Item>) => {
-      const { index, uniqueId } = itemIds(props)
-      updateOrderForm(index, { quantity: props.quantity })
-      const taskId = 'OrderItems-updateQuantity'
-      let items = [{ uniqueId, quantity: props.quantity }]
-
-      if (lastUpdateTaskRef.current.promise && isWaiting(taskId)) {
```

Comment:
yeah, this optimization is a bit tricky with the offline functionalities, but I could try to replicate it in a follow-up PR, since this one is a bit big already.